### PR TITLE
[CUDA][cuFFT] Initialize CUDA context for cuFFT before execute is called

### PR DESCRIPTION
**Pull request by eqy**
Potential fix for #109448

CC @Aidyn-A 

cc @ptrblck

### CODE REVIEW

**Review by Aidyn-A — COMMENTED** [review]
Seems like this PR does solve the issue with backward of FFT. However, I suggest not to use `cudaFree` here.

**Review by Aidyn-A — APPROVED** [review]

**Comment by ezyang** [comment]
@jithunnair-amd could you please advise for ROCm compatibility?

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 1]
```diff
@@ -304,6 +304,14 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
   CUFFT_CHECK(cufftSetWorkArea(plan, workspace.mutable_data_ptr()));
 
   // execute transform plan
+  CUcontext pctx = nullptr;
+  at::globalContext().getNVRTC().cuCtxGetCurrent(&pctx);
+  if (!pctx) {
+    // workaround for corner case where a primary context exists but is not
+    // the current context
+    at::globalContext().getNVRTC().cuDevicePrimaryCtxRetain(&pctx, 0);
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:312]: Do you mind printing a warning here (as correct me if I'm wrong almost not real app should ever get there, should they?)

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 2]
```diff
@@ -304,6 +304,14 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
   CUFFT_CHECK(cufftSetWorkArea(plan, workspace.mutable_data_ptr()));
 
   // execute transform plan
+  CUcontext pctx = nullptr;
+  at::globalContext().getNVRTC().cuCtxGetCurrent(&pctx);
+  if (!pctx) {
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:309]:
```suggestion
  if (C10_UNLIKELY(!pctx)) {
```

File: test/test_spectral_ops.py [hunk 3]
```diff
@@ -874,6 +874,22 @@ def plan_cache_max_size(device, n):
                             self.assertEqual(torch.backends.cuda.cufft_plan_cache.max_size, 10)  # default is cuda:0
                         self.assertEqual(torch.backends.cuda.cufft_plan_cache.max_size, 11)  # default is cuda:1
 
+    @deviceCountAtLeast(1)
```
Comment (malfet) [inline test/test_spectral_ops.py:877]: This is not needed, is it? (i.e. `@onlyCUDA` decorator already implies that you have a device to run the test) 
```suggestion
```

File: test/test_spectral_ops.py [hunk 4]
```diff
@@ -874,6 +874,22 @@ def plan_cache_max_size(device, n):
...
+    @onlyCUDA
+    @dtypes(torch.cfloat)
```
Comment (malfet) [inline test/test_spectral_ops.py:879]: Why not test for cdouble? Or does it work like that?
```suggestion
    @dtypes(torch.cfloat, torch.cdouble)
```

File: test/test_spectral_ops.py [hunk 5]
```diff
@@ -874,6 +874,22 @@ def plan_cache_max_size(device, n):
...
+    def test_cufft_context(self, devices, dtype):
+        x = torch.randn(32, dtype=dtype, device=devices[0], requires_grad=True)
+        dout = torch.zeros(32, dtype=dtype, device=devices[0])
```
Comment (malfet) [inline test/test_spectral_ops.py:882]: Nit
```suggestion
    def test_cufft_context(self, device, dtype):
        x = torch.randn(32, dtype=dtype, device=device, requires_grad=True)
        dout = torch.zeros(32, dtype=dtype, device=device)
```

File: test/test_spectral_ops.py [hunk 6]
```diff
@@ -874,6 +874,22 @@ def plan_cache_max_size(device, n):
                             self.assertEqual(torch.backends.cuda.cufft_plan_cache.max_size, 10)  # default is cuda:0
                         self.assertEqual(torch.backends.cuda.cufft_plan_cache.max_size, 11)  # default is cuda:1
 
+    @deviceCountAtLeast(1)
+    @onlyCUDA
+    @dtypes(torch.cfloat)
+    def test_cufft_context(self, devices, dtype):
+        x = torch.randn(32, dtype=dtype, device=devices[0], requires_grad=True)
```
Comment (malfet) [inline test/test_spectral_ops.py:881]:
```suggestion
        # Regression test for https://github.com/pytorch/pytorch/issues/109448
        x = torch.randn(32, dtype=dtype, device=devices[0], requires_grad=True)
```

**Review by malfet — APPROVED** [review]
LGTM, but please fix ROCM compilation issues

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 7]
```diff
@@ -304,6 +304,14 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
   CUFFT_CHECK(cufftSetWorkArea(plan, workspace.mutable_data_ptr()));
 
   // execute transform plan
+  CUcontext pctx = nullptr;
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:307]:
```suggestion
#if !defined(USE_ROCM)
  CUcontext pctx = nullptr;
```

File: aten/src/ATen/native/cuda/SpectralOps.cpp [hunk 8]
```diff
@@ -304,6 +304,14 @@ static const Tensor& _exec_fft(Tensor& out, const Tensor& self, IntArrayRef out_
...
+  at::globalContext().getNVRTC().cuCtxGetCurrent(&pctx);
+  if (!pctx) {
+    // workaround for corner case where a primary context exists but is not
+    // the current context
+    at::globalContext().getNVRTC().cuDevicePrimaryCtxRetain(&pctx, 0);
+    at::globalContext().getNVRTC().cuCtxSetCurrent(pctx);
+  }
```
Comment (malfet) [inline aten/src/ATen/native/cuda/SpectralOps.cpp:314]:
```suggestion
  }
#endif /* !defined(USE_ROCM) */
```

**Pull request closed without merging**